examples/wmt/input_pipeline.py

```diff
@@ -0,0 +1,255 @@
+# Copyright 2020 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Input pipeline for the wmt de-en dataset.
+
+NOTES:
+  - Not set up for dynamic batching with multihost training!
+"""
+import os
+from absl import logging
+import tensorflow.compat.v2 as tf
+import tensorflow_datasets as tfds
+
+AUTOTUNE = tf.data.experimental.AUTOTUNE
+
+
+def save_tfds_subword_vocab(encoder, path):
+  """Save tfds SubwordTextEncoder vocab to given path."""
+  subword_data = '\n'.join(["'%s'" % s for s in encoder.subwords])
+  with tf.io.gfile.GFile(path, 'w') as fp:
+    fp.write(subword_data)
+
+
+def load_tfds_subword_vocab(path):
+  """Load tfds SubwordTextEncoder vocab from given path."""
+  with tf.io.gfile.GFile(path, 'r') as fp:
+    subword_data = fp.read()
+  vocab_list = [line[1:-1] for line in subword_data.split('\n')]
+  return tfds.features.text.SubwordTextEncoder(vocab_list=vocab_list)
+
+
+def bin_and_batch(dataset,
+                  training,
+                  n_devices,
+                  batch_size=256,
+                  bucket_length=32,
+                  buckets=None,
+                  drop_remainder=True,
+                  max_eval_length=None):
+  """Batching function."""
+  # Create heuristic buckets is none are specified.
+  if buckets is None:
+    logging.info('Heuristically bucketing based on shapes of examples.')
+    bucket_boundaries = [
+        bucket_length // 4, bucket_length // 2, bucket_length,
+        bucket_length * 2, bucket_length * 4, bucket_length * 8,
+        bucket_length * 16
+    ]
+    bucket_batch_sizes = [
+        batch_size * 4, batch_size * 2, batch_size,
+        batch_size // 2, batch_size // 4, batch_size // 8,
+        batch_size // 16
+    ]
+    # allow for different evaluation max-length bucket and batchsize
+    if not training:
+      max_eval_length = max_eval_length or bucket_length * 32
+      bucket_boundaries[-1] = max_eval_length
+      bucket_batch_sizes[-1] = batch_size // max_eval_length
+    # We will pad to boundaries which pads to bucket_boundary - 1: add 1 here.
```

Comment:
What does "We will pad to boundaries" mean? I'm having a hard time parsing this sentence, would it be possible to rephrase it?

Reply:
man that's terrible english. sorry - tf.data's bucket_by_sequence_length pads to bucket_boundary - 1 so you have to add a 1 to actually set it to the desired length.  fixed comment.